Create solution for Grass Planting (#797)
cpinitiative/usaco-guide: solutions/USACO jan19-grass-plantin.mdx

---
id: usaco-894
source: USACO Silver 2019 January
title: Grass Planting
author: Aadit Ambadkar
---

> 

Reviewer: Put the link to the official editorial.

> ## Solution 1 (Smart & Intended Solution)

Reviewer 2: ```suggestion
## Solution 1 - Intended Solution
```

Don't want to imply the other is dumb :)

$\mathcal{O}(N)$ using key insight.

> Let ```adj[i]``` represent the number of paths adjacent to node $i$ in the graph. Then, considering that node, and that node alone, you would need exactly ```adj[i]+1``` numbers to fill it. This is because you would need ```adj[i]``` for each of the adjacent nodes, and $1$ for the node itself. So, our answer is the maximum of all the adjacents, plus $1$. 

Reviewer 3: i think the offical editorial explains it well enough

Author: I was requested by other people to add an editorial to my solutions. Could you please confer with them first before requesting this. Or should the rules be changed?

Reviewer 3: oh. Who may I ask? 👀

> Let ```adj[i]``` represent the number of paths adjacent to node $i$ in the graph. Then, considering that node, and that node alone, you would need exactly ```adj[i]+1``` numbers to fill it. This is because you would need ```adj[i]``` for each of the adjacent nodes, and $1$ for the node itself. So, our answer is the maximum of all the adjacents, plus 1. My Java code is below.

Reviewer 2: ```suggestion
Let ```adj[i]``` represent the number of paths adjacent to node $i$ in the graph. Then, considering that node, and that node alone, you would need exactly ```adj[i]+1``` numbers to fill it. This is because you would need ```adj[i]``` for each of the adjacent nodes, and $1$ for the node itself. So, our answer is the maximum of all the adjacents, plus $1$. 
```

For inline code you can use single `'s. for example:
```
Let `adj[i]` represent ...
```

Also when writing variables we prefer to stick with latex - in this case you would use "$\texttt{adj}[i]$" over "\`adj[i]\`"

Finally, by "adj" do you mean the "degree" or "deg"?

```java
import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class planting {
    public static void main(String[] args) throws IOException {
        BufferedReader f = new BufferedReader(new FileReader("planting.in"));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
        StringTokenizer st = new StringTokenizer(f.readLine());
        int n = Integer.parseInt(st.nextToken());
        if (n==1) {
            pw.println(1);
            pw.close();
        } else if (n==2) {
            pw.println(2);
            pw.close();

>         }

Reviewer 3: ... not necessary

        int[] adj = new int[n];

>         Arrays.fill(adj, 1);

Reviewer 3: not necessary here either

        int max=1;
        for (int i = 0; i < n-1; i++) {
            st = new StringTokenizer(f.readLine());
            int ne = Integer.parseInt(st.nextToken())-1;
            if (++adj[ne]>adj[max]) max=ne;
            ne = Integer.parseInt(st.nextToken())-1;
            if (++adj[ne]>adj[max]) max=ne;

>         }

Reviewer 3: why not an adj array but just keep track of the max value and add 1 at the end?

Author: I do think that is exactly what is happening here.

Reviewer 3: use `ans = Math.max()` and do `ans+1` at the end.

        pw.println(adj[max]);
        pw.close();
    }
}
```

> ## Solution 2 - DFS

Reviewer 3: also isnt this sol 1 but way more redundant?

Author: Which? The DFS? No, it is not really redundant. It's just slightly slower.

Reviewer 3: isn't it just counting the number of neighboring nodes? then isn't sol 1 the same as this??

> ## Solution 2 (DFS, not intended)

Reviewer 2: ```suggestion
## Solution 2 - DFS
```

This can also be solved using DFS in $\mathcal{O}(V+E)$, which for this problem is $\mathcal{O}(N)$.

```java
import java.io.*;
import java.util.*;

public class Main {

>     public static int[] visited;

Reviewer 3: why not boolean?

>     public static ArrayList<ArrayList<Integer>> adj;

Reviewer 3: Adjacency lists are usually List<Integer>[]s not two nested lists

> This can also be solved using DFS in $\mathcal{O}(V+E)=\mathcal{O}(N+N-1)=\mathcal{O}(2N-1)$, which is linear time (although slightly slower than the previous solution), enough to pass for all testcases. The implementation is straight forward. My java code for this is below:

Reviewer 2: ```suggestion
This can also be solved using DFS in $\mathcal{O}(V+E)$, which for this problem is $\mathcal{O}(N)$.
```

Unnecessary information

    public static void main(String[] args) throws IOException {
        BufferedReader f = new BufferedReader(new FileReader("planting.in"));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
        StringTokenizer st = new StringTokenizer(f.readLine());
        int farms = Integer.parseInt(st.nextToken());
        adj = new ArrayList<>();
        for (int i = 0; i < farms; i++) {
            adj.add(new ArrayList<Integer>());
        }
        for (int i = 0; i < farms - 1; i++) {
            st = new StringTokenizer(f.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            adj.get(a - 1).add(b - 1);
            adj.get(b - 1).add(a - 1);
        }
        visited = new int[farms];

>         Arrays.fill(visited, 0);

Reviewer 2: I think java initializes to 0 so this may be redundant?

Reviewer 2: This isn't really an issue -- I don't think it would cause any confusion.

Reviewer 3: it is redundant

        visited[0] = 1;
        dfs(0, 0);
        int max = 0;
        for (int i : visited) {
            max = Math.max(max, i);
        }
        pw.println(max);
        pw.close();
    }

    public static void dfs(int self, int parent) {
        int len = adj.get(self).size();
        int g = 1;
        for (int i : adj.get(self)) {
            if (i != parent) {

>                 while (g == visited[self] || g == visited[parent]) g++;

Reviewer 2: Isn't `g == visited[self]` always false?

Author: No, as g is incremented starting at 1. Supposed visited[self]=2 (val of 2). Then g=1 for the first adjacency, then g is incremented to 2, which causes a truth.

Reviewer 2: You're right -- I misread.
